src/sage/rings/lazy_species.py

```diff
@@ -0,0 +1,1395 @@
+
+    def structures(self, *labels):
+        r"""
+        Iterate over the structures on the given set of labels.
+
+        Generically, this yields a list of pairs consisting of a
+        molecular species and a relabelled representative of the
+        cosets of corresponding groups.
+
+        The relabelling is such that the first few labels correspond
+        to the first factor in the atomic decomposition, etc.
+
+        EXAMPLES::
+
+            sage: L = LazySpecies(QQ, "X")
+            sage: E = L(lambda n: SymmetricGroup(n))
+            sage: list(E.structures([1,2,3]))
+            [(E_3, ((1, 2, 3),))]
+
+            sage: P = L(lambda n: CyclicPermutationGroup(n))
+            sage: list(P.structures([1,2,3]))
+            [(C_3, ((1, 2, 3),)), (C_3, ((1, 3, 2),))]
+
+            sage: F = 1/(2-E)
+            sage: sorted(F.structures([1,2,3]))
+            [(E_3, ((1, 2, 3),)),
+             (X*E_2, ((1,), (2, 3)), 0),
+             (X*E_2, ((1,), (2, 3)), 1),
+             (X*E_2, ((2,), (1, 3)), 0),
+             (X*E_2, ((2,), (1, 3)), 1),
+             (X*E_2, ((3,), (1, 2)), 0),
+             (X*E_2, ((3,), (1, 2)), 1),
+             (X^3, ((1,), (2,), (3,))),
+             (X^3, ((1,), (3,), (2,))),
+             (X^3, ((2,), (1,), (3,))),
+             (X^3, ((2,), (3,), (1,))),
+             (X^3, ((3,), (1,), (2,))),
+             (X^3, ((3,), (2,), (1,)))]
+
+            sage: from sage.rings.species import PolynomialSpecies
+            sage: L = LazySpecies(QQ, "X, Y")
+            sage: P = PolynomialSpecies(QQ, "X, Y")
+            sage: XY = L(P(PermutationGroup([], domain=[1, 2]), {0: [1], 1: [2]}))
+            sage: list((XY).structures([1], ["a"]))
+            [(X*Y, ((1,), ('a',)))]
+
+            sage: sorted(E(XY).structures([1,2], [3, 4]))
+            [((E_2, ((((1, 'X'), (3, 'Y')), ((2, 'X'), (4, 'Y'))),)),
+              ((X*Y, ((1,), (3,))), (X*Y, ((2,), (4,))))),
+             ((E_2, ((((1, 'X'), (4, 'Y')), ((2, 'X'), (3, 'Y'))),)),
+              ((X*Y, ((1,), (4,))), (X*Y, ((2,), (3,)))))]
+
+            sage: list(XY.structures([], [1, 2]))
+            []
+        """
+        yield from self[sum(map(len, labels))].structures(*labels)
+
+    def _test_structures(self, tester=None):
+        r"""
+        Check that structures and generating series are consistent.
+
+        We check all structures on 3 labels.
+
+        TESTS::
+
+            sage: from sage.rings.species import PolynomialSpecies
+            sage: L = LazySpecies(QQ, "X, Y")
+            sage: P = PolynomialSpecies(QQ, "X, Y")
+            sage: XY = L(P(PermutationGroup([], domain=[1, 2, 3]), {0: [1], 1: [2, 3]}))
+            sage: XY._test_structures()
+        """
+        n = 3
+        P = self.parent()
+        if P._arity == 1:
+            labels = list(range(n))
+            s = list(self.structures(labels))
+            assert len(s) == len(set(s)), f"structures for {labels} are {s}, which is not a set"
```

Comment:
```suggestion
            tester.assertEqual(len(s), len(set(s)), f"structures for {labels} are {s}, which is not a set")
```
and similar such changes.